components/net/http_cache.rs

```diff
@@ -563,6 +587,151 @@ impl HttpCache {
+        if let Some(Ok(location)) = response
+            .headers
+            .get(header::LOCATION)
+            .map(HeaderValue::to_str)
+        {
+            if let Ok(url) = request.current_url().join(location) {
+                self.invalidate_for_url(&url);
+            }
+        }
+        if let Some(Ok(ref content_location)) = response
+            .headers
+            .get(header::CONTENT_LOCATION)
+            .map(HeaderValue::to_str)
+        {
+            if let Ok(url) = request.current_url().join(&content_location) {
+                self.invalidate_for_url(&url);
+            }
+        }
+        self.invalidate_for_url(&request.url());
+    }
+}
+
+fn check_vary_headers(request: &Request, cached_resource: &CachedResource) -> bool {
+    let mut can_be_constructed = true;
+    let cached_headers = cached_resource.data.metadata.headers.lock().unwrap();
+    let original_request_headers = cached_resource.request_headers.lock().unwrap();
+    if let Some(vary_value) = cached_headers.typed_get::<Vary>() {
+        if vary_value.is_any() {
+            can_be_constructed = false
+        } else {
+            // For every header name found in the Vary header of the stored response.
+            // Calculating Secondary Keys with Vary <https://tools.ietf.org/html/rfc7234#section-4.1>
+            for vary_val in vary_value.iter_strs() {
+                match request.headers.get(vary_val) {
+                    Some(header_data) => {
+                        // If the header is present in the request.
+                        if let Some(original_header_data) = original_request_headers.get(vary_val) {
+                            // Check that the value of the nominated header field,
+                            // in the original request, matches the value in the current request.
+                            if original_header_data != header_data {
+                                can_be_constructed = false;
+                                break;
+                            }
+                        }
+                    },
+                    None => {
+                        // If a header field is absent from a request,
+                        // it can only match a stored response if those headers,
+                        // were also absent in the original request.
+                        can_be_constructed = original_request_headers.get(vary_val).is_none();
+                    },
+                }
+                if !can_be_constructed {
+                    break;
+                }
+            }
+        }
+    }
+    can_be_constructed
+}
+
+/// The various states a HttpCacheEntry can be in.
+#[derive(Debug, Eq, PartialEq)]
+enum CacheEntryState {
+    /// The entry is fully up-to-date,
+    /// there are no pending concurrent stores,
+    /// and it is ready to construct cached responses.
+    ReadyToConstruct,
+    /// The entry is pending a concurrent store.
+    PendingStore,
```

Comment:
If we did want to treat the first read of a prefetch different;ly, we'd probably do it by adding other states.

Reply:
I'm starting to think that we might want to have some sort of aggressive in memory "prefecth cache" living in script versus as adapting the one in net(beyond the current changes, which are generally useful). 

Because in fact most prefetching will be wasted, unless the response contains appropriate headers. So by having a prefetch cache sitting in front of the network, and retain basically every `GET` responses(but let's say only until the document has loaded), you can be sure every prefetch actually is used to make the page load faster.  Some of those prefetches would then also populate the http-cache, if their response contains appropriate headers.

Reply:
Hmm, that might not be such a bad idea. It would stop the prefetch logic from polluting the http cache, otherwise we're in danger of accumulating complexity.

Reply:
Ok I've opened an issue at https://github.com/servo/servo/issues/24231